ext/standard/streamsfuncs.c

```diff
@@ -756,115 +782,132 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
+        if (set_count > max_set_count)
+            max_set_count = set_count;
+        sets += set_count;
+    }
 
-	if (!sets) {
-		zend_value_error("No stream arrays were passed");
-		RETURN_THROWS();
-	}
+    if (e_array != NULL) {
+        set_count = stream_array_to_fd_bigset(e_array, &efds, &max_fd);
+        if (set_count > max_set_count)
+            max_set_count = set_count;
+        sets += set_count;
+    }
 
-	if (!PHP_SAFE_MAX_FD(max_fd, max_set_count)) {
-		RETURN_FALSE;
-	}
+    if (!sets) {
+        zend_value_error("No stream arrays were passed");
+        FD_BIGSET_FREE(&rfds);
+        FD_BIGSET_FREE(&wfds);
+        FD_BIGSET_FREE(&efds);
+        RETURN_THROWS();
+    }
 
-	if (secnull && !usecnull) {
-		if (usec != 0) {
-			zend_argument_value_error(5, "must be null when argument #4 ($seconds) is null");
-			RETURN_THROWS();
-		}
-	}
+    if (secnull && !usecnull) {
+        if (usec != 0) {
+            zend_argument_value_error(5, "must be null when argument #4 ($seconds) is null");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        }
+    }
 
-	/* If seconds is not set to null, build the timeval, else we wait indefinitely */
-	if (!secnull) {
-		if (sec < 0) {
-			zend_argument_value_error(4, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		} else if (usec < 0) {
-			zend_argument_value_error(5, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		}
+    if (!secnull) {
+        if (sec < 0) {
+            zend_argument_value_error(4, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        } else if (usec < 0) {
+            zend_argument_value_error(5, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        }
+
+        tv.tv_sec = (long)(sec + (usec / 1000000));
+        tv.tv_usec = (long)(usec % 1000000);
+        tv_p = &tv;
+    }
 
-		/* Windows, Solaris and BSD do not like microsecond values which are >= 1 sec */
-		tv.tv_sec = (long)(sec + (usec / 1000000));
-		tv.tv_usec = (long)(usec % 1000000);
-		tv_p = &tv;
-	}
-
-	/* slight hack to support buffered data; if there is data sitting in the
-	 * read buffer of any of the streams in the read array, let's pretend
-	 * that we selected, but return only the readable sockets */
-	if (r_array != NULL) {
-		retval = stream_array_emulate_read_fd_set(r_array);
-		if (retval > 0) {
-			if (w_array != NULL) {
-				zval_ptr_dtor(w_array);
-				ZVAL_EMPTY_ARRAY(w_array);
-			}
-			if (e_array != NULL) {
-				zval_ptr_dtor(e_array);
-				ZVAL_EMPTY_ARRAY(e_array);
-			}
-			RETURN_LONG(retval);
-		}
-	}
+    // Support for buffered data
```

Comment:
Please refrain from changing / removing comments unless they are incorrect or related to your changes